### DISCUSSION TITLE
v1.3.1 and dbus use

### DISCUSSION THREAD

**Discussion by 0323pin** [tsujan/FeatherPad#670] [General]
Hi,

NetBSD package maintainer here and long time (3 years) `featherpad` user.

As I understand from the ChangeLog, the application now uses `dbus` but, does not actually require it to function.

So, I'm guessing it will detect it's presence and use it when available. Will this always be the case or, should I consider making it a requirement at build time already now?

Thanks for any hints.

**Comment by tsujan** [comment]
> Will this always be the case?

Yes. It was tested under macOS without D-Bus (Thanks to @shlyakpavel), and it shouldn't be different under NetBSD: If D-Bus exists, FeatherPad will have a single process; otherwise, each FeatherPad window will have its own process — nothing essential will be lost.

> should I consider making it a requirement at build time already now?

You probably mean a *runtime* dependency. It isn't necessary (for the above-mentioned reason) but possible.

**Reply by 0323pin** [reply]
Thanks!
I do have `dbus` on my system because it's a `firefox` dependency and I'm using LeftWM as my WM which requires a `dbus` session.
Yes, I meant a runtime dependency but, I won't do it if it's not needed.
Btw, I've already merged the update anyway :)
https://mail-index.netbsd.org/pkgsrc-changes/2022/08/04/msg257968.html

**Reply by tsujan** [reply]
> Btw, I've already merged the update anyway

Thanks!

> but, I won't do it if it's not needed.

No, it isn't essential.

In all my apps, I avoid backward-incompatible changes as far as possible. I don't remember a case, where backward incompatibility was needed. But I don't deny that it might be required in other apps — or even in my apps at some point.

**Comment by 0323pin** [comment]
@tsujan keep-up the good work, thanks!

**Reply by tsujan** [reply]
Thanks for the kind words!

**Comment by 0323pin** [comment]
@tsujan Out-of-topic here but, I thought I'd ask. Would you mind having a look at https://www.unitedbsd.com/d/888-qt-applications-cant-open-files-which-have-special-characters-in-their-name and linked question https://www.unitedbsd.com/d/606-keyboard-dead-keys-dont-work-in-qt-applications

Anyway to get featherpad to recognize special characters when used standalone?
Thought, I'd ask just in case I'm missing something obvious.

Thanks!

**Reply by tsujan** [reply]
That can't be related to FeatherPad, its standalone mode, or LXQt file dialog. It should be about something much more basic.

I don't know about *BSD, but in Linux, all file names are in UTF-8, which supports all possible characters in the world.

Of course, the font you use may not support all characters. For Korean, Japanese, Chinese and some other languages, you need appropriate fonts. Since I installed them a long time ago, I don't remember their names; sorry!

However, something fundamental may be missing or wrong in that system, because a simple text like "caffè" shouldn't make any trouble with any well-known font like DejaVu Sans. Actually, if the name was in Korean, Chinese, Japanese,..., it shouldn't make any trouble in file opening either; at most, in the absence of appropriate fonts, file names wouldn't be human-readable.

**Reply by 0323pin** [reply]
Thank you!

**Reply by tsujan** [reply]
https://www.unitedbsd.com/d/606-keyboard-dead-keys-dont-work-in-qt-applications/ suggests that there's a general problem with Qt. Perhaps it's related to how Qt is compiled there (just a wild guess). Ping me if you find the cause; I got curious.

**Reply by 0323pin** [reply]
Yes, I know. This is what I'm trying to understand since, Gtk-apps don't have this issue and recognize the characters.

**Comment by 0323pin** [comment]
Btw, I've quoted your answers here on the forum.
Let me know if you don't want that and I'll remove it.

I have a feeling I need to dig deep to find out why, if I manage to find it 😞

**Reply by tsujan** [reply]
Because Qt has never had that problem here, and supposing that it happens anywhere Qt apps are used, logically, I think there are three possibilities: (1) Something may be wrong in Qt for BSD (if it exits at all); (2) It may be related to how Qt is compiled/used; or (3) The cause may be outside Qt but related to it.

If it doesn't happen everywhere Qt apps are used, then the problem is definitely outside Qt.

But my knowledge of BSD is next to zero; so, please don't be misled by my words!

> Let me know if you don't want that and I'll remove it.

I have no problem with quotations at all.

**Comment by 0323pin** [comment]
> (2) It may be related to how Qt is compiled/used

Yes, it happens with all Qt apps, OP mentioned the same issue with strawberry, and your (2) is what I suspect. I did ask another NetBSD dev on that thread but, got nothing back. Quoting myself,

>  I've just added inputmethod/uim to the lxqt meta-package for now. I didn't look at it anyfurther but, one thing pops-up. Why is this not the case with Gtk?

I'll go through the Qt packages to see if I find something obvious in the compile options.

**Comment by 0323pin** [comment]
@tsujan OP on the thread came back with the following observation,

> It can't be a bug in FeatherPad specifically because I get the same with Strawberry. However it isn't a general issue with Qt programs either: PCManFM-Qt and qpdfview aren't affected.

Indeed, it's true 😕 

`pcmanfm-qt` and `featherpad` file-chooser,
![2022-08-16-082552_1366x768_scrot](https://user-images.githubusercontent.com/90570748/184813754-da68ad52-7e47-4f64-b21e-210b48cebd81.png)

If I add the `.pdf` file extension is also recognized by `xpdf` file-chooser,
![2022-08-16-082642_1366x768_scrot](https://user-images.githubusercontent.com/90570748/184813924-f8b8bdaa-1b8c-495a-ae1f-e11b9a24204f.png)

So, now I'm going with your option (3)

> (3) The cause may be outside Qt but related to it.

But, this will make it harder to find the cause for it 😞 
Now, I remember why I've just added `uim` input method to the LXQt meta-package.

**EDIT:** I'm only getting more confused, see the new posts on the thread, https://www.unitedbsd.com/d/888-qt-applications-cant-open-files-which-have-special-characters-in-their-name/11

**Reply by tsujan** [reply]
> I'm only getting more confused

Let me explain what a native file dialog is.

Qt comes with a simple file dialog, like this (ignore the style; it's Kvantum):

![qt_file_dialog](https://user-images.githubusercontent.com/981076/184887848-05b956e8-194f-4144-afc8-c393932c5b8e.png)

Although it can be used, it lacks many features. If you're inside a DE without Qt support, you'll see that dialog. It's also called Qt's "default" dialog.

Qt-based DEs, like LXQt and KDE, implement their rich file dialogs. Qt has an efficient plugin system that allows that. This is LXQt file dialog:

![lxqt_file_dialog](https://user-images.githubusercontent.com/981076/184888554-e36bcbb0-2e91-48e5-b05c-4a042c5cf130.png)

It's called "native" under LXQt, as KDE file dialog is called so under KDE.

FeatherPad and other Qt apps don't have file dialogs of their own. They either use the native file dialog if it exists or Qt's default file dialog if it doesn't. FeatherPad has an extra option for that, which is checked by default ("Native file dialog").

**Reply by 0323pin** [reply]
Thanks for the concept explanation.
The thread OP is using LXQt. I'm using `featherpad`, `pcmanfm-qt` and `xpdf` all outside a Qt-based DE, just `LeftWM` as the window manager.

As you can see in the screenshots, both `pcmanfm-qt` and `xpdf` using the simple Qt file-dialog show the file with the special characters.
`featherpad` does not and, according to OP, it's the same with `strawberry`.

What I'm trying to understand is where this difference comes from, so I can change or modify the builds in order to produce binaries that are consistent across applications.

**EDIT:** There's nothing wrong with `featherpad`, it works on Void Linux. I might need to change something in our build, though.

**Reply by tsujan** [reply]
> As you can see in the screenshots, both pcmanfm-qt and xpdf using the simple Qt file-dialog show the file with the special characters.
featherpad does not...

It's Qt's default file dialog, which is called by FeatherPad.

Yes, that's a mystery to me too: why doesn't Qt's default file dialog show "caffè" there? Of course, it does here.

**Reply by tsujan** [reply]
A screenshot:

![caffe](https://user-images.githubusercontent.com/981076/184900369-7fe84bf1-4544-4ef0-ab8e-cdb3bdc54235.png)

**Reply by 0323pin** [reply]
Yes, as mentioned it works on Void Linux also, both with `xpdf` and `featherpad`. On my machine it works with `xpdf` but not with `featherpad` 😩 

It should be no difference, as it's the same file-dialog which is being used by both applications.

Just to be on the safe side here is the list of runtime dependencies I've specified for the build of the binary package,

```
.include "../../x11/qt5-qtbase/buildlink3.mk"
.include "../../x11/qt5-qtx11extras/buildlink3.mk"
.include "../../x11/qt5-qtsvg/buildlink3.mk"
.include "../../textproc/hunspell/buildlink3.mk"
.include "../../x11/libX11/buildlink3.mk"
.include "../../x11/libXext/buildlink3.mk"
.include "../../graphics/hicolor-icon-theme/buildlink3.mk"
```

Besides these, it has buildtime dependencies on `pkg-config`, `qt5-qttools` and `cmake`

**Reply by pengxingwei** [reply]
> A screenshot:
> 
> ![caffe](https://user-images.githubusercontent.com/981076/184900369-7fe84bf1-4544-4ef0-ab8e-cdb3bdc54235.png)

Can you please tell me the name of lxqt theme and icon theme as shown in picture?

**Reply by tsujan** [reply]
> ...the name of lxqt theme and icon theme as shown in picture?

There's nothing specific to LXQt in the screenshot. The Qt style is a [Kvantum](https://github.com/tsujan/Kvantum) theme (Kvantum Manager → Change/Delete Theme → Select a theme: → KvMojaveMixed1). The Icon set is [Plataro](https://github.com/tsujan/Plataro). Both are DE-independent. The window decoration is [BreezeEnhanced](https://github.com/tsujan/BreezeEnhanced) (for KWin, here used under LXQt).

**Comment by tsujan** [comment]
A very wild guess. Please ignore it if you feel it's misleading; I don't want to create more confusion, especially when I don't know anything about BSD.

In Linux, all file names are in UTF-8. *Maybe*, in BSD, they aren't. *Maybe*, when you name a file as "caffè" in BSD, it isn't in UTF-8 (unless you somehow tell BSD to make it so).

pcmanfm-qt and LXQt file dialog can deal with non-UTF-8 file names because they use `GLib` for file managemnet; Qt's default file dialog can't.

**Reply by 0323pin** [reply]
You could be on to something here but, as I wrote above (posted after this last post of yours), `xpdf` is also using the basic file-dialog from Qt. It might be something with the dependencies.

**Reply by tsujan** [reply]
What happens if you give the path to FeatherPad directly in a terminal: `featherpad /home/pin/caffè`?

**Reply by 0323pin** [reply]
It opens a `featherpad` window with a warning `The file does not exist`. Interesting enough the name displays correctly in the tab.
It only gets weird.

**Reply by tsujan** [reply]
> nteresting enough the name displays correctly in the tab.

That's normal. The name should be in the tab, in case the user wants to save a file that doesn't exist.

> It opens a featherpad window with a warning `The file does not exist`.

That confirms my theory: `/home/pin/caffè` doesn't exist, and the `caffè` you see isn't the UTF-8 `caffè` in the command-line.

**Reply by 0323pin** [reply]
Ok! Obviously I can't type these characters inside `featherpad` or `feathernotes` for that matter.
I need to find how to fix this. Did you see user rvp analysis on the thread?

Quoting,
====

A `ktrace` shows what's happening:
```sh
$ ktrace fpad
$ kdump > k.txt
[...]
  1051     31 fpad     CALL  __lstat50(0x7aef1933be00,0x7aeef6fffcc0)
  1051     31 fpad     NAMI  "/mnt/snd/Korean/??? ?? ??-dNyDsJtY2sI.webm"
  1051     31 fpad     RET   __lstat50 -1 errno 2 No such file or directory
[...]
```
Something is converting UTF-8 strings into (probably) US-ASCII using the standard [iconv()](https://man.netbsd.org/NetBSD-9.3/amd64/iconv.3) function and then passing _this_ string (instead of the original filename) when opening the file. (`iconv()` replaces any valid multibyte sequence it can't convert with a `?` char.)

In the "native" file chooser, the filenames are passed correctly when listing the directory, however:
```text
[...]
  1051     28 fpad     CALL  access(0x7aef187c6180,2)
  1051     28 fpad     NAMI  "/mnt/snd/Korean/아이유 너의 의미-dNyDsJtY2sI.webm"
  1051     28 fpad     RET   access -1 errno 30 Read-only file system
[...]
```
You can do this same conversion yourself:
```sh
$ ls /mnt/snd/Korean/ | iconv -s -f utf8 -t us-ascii
Shin Joong Hyun & The Men - Beautiful Rivers and Mountains-d_3Qe7KfBJA.webm
playlist.m3u
???? ???? The Meaning of You-zXkhG2rSkAo.webm
??? ?? ??-dNyDsJtY2sI.webm
$
```
BTW, Featherpad 1.2.0 on FreeBSD shows exactly the same behaviour. I can't install the program on my Ubuntu 19.04 'cuz it's too old and the repository has disappeared.

Someone who knows LXQT/QT ought to investigate...

**Reply by tsujan** [reply]
So, it *is* about encoding. I wonder why BSD should use non-UTF-8 encoding. Nowadays, even MS Windows uses UTF-8.

**Reply by tsujan** [reply]
> Obviously I can't type these characters inside featherpad or feathernotes for that matter.

Qt widgets only get and show UTF-8 strings. FeatherPad can show non-UTF-8 text files only because it has codes for guessing their encodings and converting them to UTF-8.

BTW, Qt6 has removed all support for legacy encodings.

**Reply by 0323pin** [reply]
Thanks for all your feedback.

It still doesn't explain why when using `xpdf` the characters are visible. I need to try a few more settings.

> Qt6 has removed all support for legacy encodings

Don't mention it. We are late with it, the packages are still in the testing repository and it's moving slow.

**Reply by tsujan** [reply]
Something like with pcmanfm-qt. Being one of its devs, I know that `GLib` works with non-UTF-8 strings. pcmanfm-qt uses `GLib` in its core code but Qt for its skin. `GLib` gets the file name and gives a UTF-8 string to Qt.

Probably, `xpdf` does that in another way.

If you can't type into FeatherPad, you shouldn't be able to type inside any Qt widget that accepts texts, like any text entry. Would you test it?

**Reply by 0323pin** [reply]
Probably, yes.

> If you can't type into FeatherPad, you shouldn't be able to type inside any Qt widget that accepts texts, like any text entry. Would you test it?

That's why I've added `uim` to the LXQt meta-pkg, to make dead keys work. I can intall `lxqt-runner` tomorrow and try but, by now I'm expecting it to fail.

Thanks for your time.

**Reply by tsujan** [reply]
> I can intall lxqt-runner

No need to it. Any text entry in any Qt program can be used. However, if you had to use `uim`, your system shouldn't be in UTF-8. 

In Linux, we have the `locale` command. Here, its output is:
```
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

**Reply by 0323pin** [reply]
Here's mine

```
~ > locale
LANG="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_COLLATE="C"
LC_TIME="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_ALL=""
```

And here's the one from Nia in the thread that has it working fine,

```
$ locale
LANG=""
LC_CTYPE="en_US.UTF-8"
LC_COLLATE="C"
LC_TIME="en_GB.UTF-8"
LC_NUMERIC="C"
LC_MONETARY="C"
LC_MESSAGES="C"
LC_ALL=""
```

**Reply by tsujan** [reply]
Actually, *yours* seems correct to me.

We're missing something about locale and UTF-8 in BSD.

**Reply by 0323pin** [reply]
@tsujan Obviously, even if Nia's output of `locale` above looked is weird, I had to try changing my settings to match hers.
After all, she said it was working for her.

Surprise, now I can see, open and type accented characters.

**Reply by tsujan** [reply]
I have no explanation, other than saying it should be related to how locale is handled by BSD. At least, the situation isn't as mysteriious as it was before ;)

**Reply by 0323pin** [reply]
Correct. I wasn't expecting you to have an explanation but, after this long conversation, I felt the least I could do was letting you know it solved.

Even if the `locale` output feels wrong.

**Reply by tsujan** [reply]
> I felt the least I could do was letting you know it solved.

Thanks!

> Even if the locale output feels wrong.

It isn't wrong, but I expected your previous locale to work better, because of `LANG="en_US.UTF-8"`.

A month ago, locale became a mess here (Manjaro) after an ugrade. I fixed the problem by reversing the weird changes they had made to `/etc/locale.gen` and `/etc/locale.conf`. A day later, Manjaro reversed them too. Apart from that, I haven't had any problem with locale.

**Reply by 0323pin** [reply]
Thankfully, my other system runs on Void with musl libc, where locale is a bit on an alien concept. English, or basically nothing, so not much to mess with 😀

**Reply by rajeevvp** [reply]
> So, it _is_ about encoding. I wonder why BSD should use non-UTF-8 encoding. Nowadays, even MS Windows uses UTF-8.

If it was copied from an UTF-8 source, the characters should be UTF-8 also. And, in all the tests I've done (see the `ktrace` stuff just above), the filesnames were all UTF-8. BSD, just like the other Unixes (incl. Linux) doesn't care what encoding filenames use (at least on the native filesystems). The only characters you can't have in any Unix filename are `\0` and `/`.

I think what's happening here is that either a) the locale is not being switched from the default `C` locale, or b) is being reset to the `C` locate. The `???` char. replacement is very characteristic of this.

If you can tell me why the test code in [this](https://www.unitedbsd.com/d/888-qt-applications-cant-open-files-which-have-special-characters-in-their-name/30) post behaves correctly, but, a similar  change in the patch doesn't, then we would be well on the way to fixing this issue.

Thx.

**Reply by tsujan** [reply]
> ... just like the other Unixes (incl. Linux) doesn't care what encoding filenames use (at least on the native filesystems).

Correct, in a sense, but practically, Linux distros are set up is such a way that file names/paths are in UTF-8. Also, I found https://why-openbsd.rocks/fact/utf8/ ("OpenBSD removed support for all non-UTF-8 locales in 2015...").

**Reply by tsujan** [reply]
I know there may be no need to emphasize the importance of UTF-8 here, but a short comment:

As https://wiki.qt.io/Strings_and_encodings_in_Qt recommends: "In your code use QStringLiteral / QString::fromUtf8() for all string literals...". That includes file paths.

And as mentioned above, Qt6 removed all support for legacy encodings. What remains consists of UTF-8/16/32 and Latin1. If you compile FeatherPad against Qt6, the menu-items for other encodings disappear from Options → Encoding.

`Glib` also uses UTF-8 for its strings.

UTF-8 covers all natural languages. It might have room for extraterrestrial languages too — if not, no worries; we have UTF-16 ;)

**Reply by rajeevvp** [reply]
EDIT: change define from `1` to `-1` to match FreeBSD.

Pin, can you apply this patch and rebuild Qt5?

```diff
--- pkgsrc.orig/x11/qt5-qtbase/Makefile.common	2022-04-19 01:01:12.000000000 +0000
+++ pkgsrc/x11/qt5-qtbase/Makefile.common	2022-08-20 07:56:01.481446492 +0000
@@ -48,7 +48,7 @@
 CFLAGS+=		-Wno-register
 
 # wcstof(3) etc. is used in C++ code.
-CFLAGS.NetBSD+=		-D_NETBSD_SOURCE
+CFLAGS.NetBSD+=		-D_NETBSD_SOURCE -D_XOPEN_UNIX=-1
 
 # include/openssl/e_os2.h:283:25: error: '_Noreturn' does not name a type
 CXXFLAGS.SunOS+=	-D_Noreturn=""
```

NetBSD doesn't define `_XOPEN_UNIX` in its header (FreeBSD defines it as `-1`) and that definition is used in various places in the Qt5 source, including:

```
qtbase-everywhere-src-5.15.5/src/corelib/codecs/qtextcodec.cpp
qtbase-everywhere-src-5.15.5/src/corelib/codecs/qiconvcodec.cpp
```

where it is used to determine the currently active locale. Since that isn't defined, the code then plays guessing games with the **opaque** string returned by `setlocale()`--which is is _highly_ system-dependent (even on Linux the results can vary wildly depending on which env. vars. are set)--and then falls back to using `ISO 8859-1` as the default codeset.

As to whether NetBSD should define these `_XOPEN_XXX` constants in `unistd.h` like Linux and FreeBSD do, or whether the Qt code should be changed to check directly for the functions which it needs (eg. `#ifdef HAVE_NL_LANGINFO` after probing it via `cmake` etc.) instead of relying on proxies like `_XOPEN_UNIX`, I don't know. That's something you NetBSD folks can decide; my preference is for the latter.

**Reply by 0323pin** [reply]
@rajeevvp yes, I can patch qtbase but not before Monday.

> As to whether NetBSD should define these _XOPEN_XXX constants in unistd.h like Linux and FreeBSD do, ... I don't know. That's something you NetBSD folks can decide

Can you please open the discussion on the mailing lists? Probably in tech-pkg after I test the patch.

**Reply by 0323pin** [reply]
@rajeevvp have you tested the patch yourself?

**Reply by 0323pin** [reply]
@rajeevvp Ops, life came in between ... I won't have time before midweek, earliest. Sorry 😞

**Reply by rajeevvp** [reply]
> have you tested the patch yourself?

Sorry, no, mate. Had stuff to do on the weekend myself. Will test in the next few days.


> As https://wiki.qt.io/Strings_and_encodings_in_Qt recommends: "In your code use QStringLiteral / QString::fromUtf8() for all string literals...". That includes file paths.
> 

That's part of the problem here. Converting filenames from UTF-8 is fine when _displaying_ them in dialog boxes and error messages, etc., but, when passing to system calls filenames should be the original stuff got from the system. Otherwise, you get the program trying to open files literally named `???.ext` depending on locale (this can be seen in the `ktrace` output above). Contrast with, say, `vi` (which ever flavour you prefer), which should open arbitrarily-encoded filenames and no matter what the user's own locale is set to.

Also note that all POSIX programs start in the `C` locale and _stay in that locale_ until the program (or a library it calls) changes the settings with something like `setlocale(LC_ALL, "");` (This sets up the program to use whatever locale stuff the user set it to via the `LC_*` env. vars, and is typically the first function call made by a [locale-aware](http://cvsweb.netbsd.org/bsdweb.cgi/src/bin/ls/ls.c?rev=1.77&content-type=text/x-cvsweb-markup&only_with_tag=MAIN) C program. I've tried to do this the C++/Qt way in the test patch in my previous post but the other bug seems to have got in the way of it working.)

**Comment by 0323pin** [comment]
@tsujan Sometimes it takes longer than it should and in the meantime FeatherPad is Qt6 based but, @rajeevvp has found the issue,
https://www.unitedbsd.com/d/606-keyboard-dead-keys-dont-work-in-qt-applications/37

It turns out there's a path missing in our `libxkbcommon` package.

**Reply by tsujan** [reply]
> It turns out there's a path missing in our libxkbcommon package.

Please also note that FeatherPad isn't limited to X11, although it has extra features under X11, which can't exist under Wayland (for now).

**Comment by 0323pin** [comment]
It's fixed now.

As for Wayland, we have a very long way to go. There are simply too many Linux only things that need to be addressed.
I've tested Wayland on NetBSD a few years ago and it was far from daily drive ready.

**Reply by tsujan** [reply]
Although it makes no difference for FeatherPad which Wayland compositor is used, Wayland isn't a unified concept; your experience of it depends on which Wayland compositor you try. Also, during last years, many things have changed in those compositors that are based on `wlroots`. I recommend LabWC.

**Comment by 0323pin** [comment]
> Also, during last years, many things have changed in those compositors that are based on wlroots. I recommend LabWC.

Nothing based on `wlroots` will currently run on NetBSD. Only [swc based](https://github.com/michaelforney/swc) compositors are supported.

**Reply by tsujan** [reply]
> Nothing based on wlroots will currently run on NetBSD.

That's OK as far as FeatherPad is concerned; its X11 support will be continued indefinitely. I just wanted to share a part of my experience about Wayland.

**Reply by 0323pin** [reply]
Appreciated. Even on Linux, I'm still running X11 only, so any insights into Wayland are welcome.

**Reply by tsujan** [reply]
> Even on Linux, I'm still running X11 only

A wise choice, IMO.

I use Wayland because I need to do coding for and under it. I also test others' works on some Wayland compositors (mainly, LabWC, Wayfire and kwin_wayland).
